### PR TITLE
upgrade of YAMTL solutions (YAMTL => 0.1.5) and new solution with explicit incrementality

### PR DESCRIPTION
**Pull request by arturboronat**
New solutions with YAMTL 0.1.5:

* Solutions now bypass the EMF Change model, which was creating some dependency with respect to the size of the model when propagating changes.
* A new variant with explicit incrementality to avoid recomputation of the query during incremental evaluation: YAMTL exposes changes to the queries, which add logic for managing those changes. The solution to Q2 is now incremental and uses an adaptation of Sedgewick and Wayne's weighted quick union-find algorithm.
* Since we agreed that memory measurements are not relevant for the article, may I suggest to disable sending hints to the garbage collector. This slows down the overall execution of the benchmark and it may interfere with performance results. These hints have been disabled in YAMTL's solutions.

The consequences of this pull request are that performance results are now improved and change propagation is performed in constant time with respect to the size of the model. Change propagation is only dependent on the size of the change.

### CODE REVIEW

**Comment by marci543** [comment]
Thank you for the PR.

>- Since we agreed that memory measurements are not relevant for the article, may I suggest to disable sending hints to the garbage collector. This slows down the overall execution of the benchmark and it may interfere with performance results. These hints have been disabled in YAMTL's solutions.

I would not remove this capabality. If I'm right, this doesn't interfere with the runtime results since the GC is invoked after the time measurements.
https://github.com/TransformationToolContest/ttc2018liveContest/blob/cd5c9686f4810d425a33271d98c915dfb201ec2a/solutions/EMFSolutionTemplate/src/ttc2018/LiveContestDriver.java#L121-L128

What do you think about this?

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,10 @@ changes/1024/
 # IntelliJ
 .idea/
 *.iml
+models-yamtl/
+solutions/EMFSolutionYAMTL_element/
+solutions/EMFSolutionYAMTL_helpersOnly_v1/
+solutions/EMFSolutionYAMTL_indexed
+solutions/EMFSolutionYAMTL_uf
+**/*.documented.xmi
+models/1024/
```
Comment (marci543) [inline .gitignore]: It is intentional to have the .gitignore in models/1024 folder to make sure that the folder exists.
Can I put it back?

**Comment by marci543** [comment]
The CI uses a cached `model` folder from the `common` base image from Docker Hub, so I'll update that manually.
What was the change in `1024.zip`? Only the removal of `*.documented.xmi`s?

**Comment by arturboronat** [comment]
It did seem to interfere in the performance. For some reason the benchmark executed much slowlier and it added a few millis to the execution of updates. Taking into account the current performance times, that interference may well be 20-30% of the time.

I prefer to keep it commented.

Artur

**Comment by arturboronat** [comment]
This folder should be ignored when making commits. I don’t mind where this .gitignore is place. Otherwise we will keep committing the whole folder.

Artur

**Comment by arturboronat** [comment]
Yes, *.documented.xmi’s have been removed. That’s the only change.

Artur

File: solutions/EMFSolutionYAMTL/solution.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [build]
-default=./gradlew clean fatJar
+default=./gradlew --refresh-dependencies clean fatJar
```
Comment (marci543) [inline solutions/EMFSolutionYAMTL/solution.ini:2]: Is the `--refresh-dependencies` option needed by other "tools", e.g. batch and expl?
Should it be run for the variant that skip tests?

**Comment by marci543** [comment]
>I prefer to keep it commented.

OK. In this case I must comment the GC call out from all other tools and rerun the whole benchmark.
Should I also comment out the print line to avoid reporting uncertain numbers?

**Comment by arturboronat** [comment]
Many thanks. This makes sense to me as we are not analysing memory consumption. I would inform other authors on Friday in any case, just in case there are other views.

On the other hand, I’m now writing the updates on the article and I think there is no point in presenting two solutions to Q2. I will build on the union-find solution and present it from three different perspectives (batch, implicit traceability and explicit traceability). This will help the comparison with other incremental solutions that also use union find, addressing the heterogeneity concern in the article reviews.

This also means that I will send another pull request by the end of tomorrow, if I find the time.

Cheers,
Artur

**Comment by marci543** [comment]
OK. Then I'll wait for that PR before rerunning the experiments.

File: solutions/EMFSolutionYAMTL/solution.ini (hunk 2 above)
Comment (marci543) [reply solutions/EMFSolutionYAMTL/solution.ini:2]: Should I put `--refresh-dependencies` option to all Gradle builds of YAMTL like [here](https://github.com/TransformationToolContest/ttc2018liveContest/pull/107#discussion_r631438170)?

File: solutions/EMFSolutionAOF/.project [hunk 3]
```diff
@@ -5,6 +5,11 @@
 	<projects>
 	</projects>
 	<buildSpec>
+		<buildCommand>
+			<name>org.eclipse.xtext.ui.shared.xtextBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
```
Comment (marci543) [inline solutions/EMFSolutionAOF/.project]: Are these changes to the EMFSolutionAOF .project file intentional? Should we keep them?

**Comment by marci543** [comment]
What do you think about the changes?

File: solutions/EMFSolutionYAMTL/solution.ini (hunk 2 above)
Comment (arturboronat) [reply solutions/EMFSolutionYAMTL/solution.ini:2]: This is only needed in the first YAMTL solution that is executed to make sure the latest snapshot for the stated version is downloaded in the .m2 repository, skipping the cached dependency. The other two subsequent solutions will use the cached version.

File: solutions/EMFSolutionAOF/.project (hunk 3 above)
Comment (arturboronat) [reply solutions/EMFSolutionAOF/.project]: No, they are not. Please ignore them.

**Comment by arturboronat** [comment]
The new commits that have been added to this PR by default contain the simplified solutions. 

Now all YAMTL solutions for the Q2 challenge are based on the weighted quick UF algorithm.

All done from my side.

Many thanks.

**Comment by arturboronat** [comment]
> What do you think about the changes?
> 
I fetched your changes and my solution runs okay. So changes seem fine on my side.

Many thanks.

File: solutions/EMFSolutionYAMTL/solution.ini (hunk 2 above)
Comment (marci543) [reply solutions/EMFSolutionYAMTL/solution.ini:2]: Fine. I'll add this also to the skipTests variant for the first solution.

**Pull request merged**